diff --git a/dist/index.js b/dist/index.js
index bc6b4e91faa3ea3f01710e479759cbb19a764698..c33d75428fb0175ee1617c91b9923c112219537e 100644
--- a/dist/index.js
+++ b/dist/index.js
@@ -2018,7 +2018,7 @@ const buildCfStyles = (values) => {
         lineHeight: values.cfLineHeight,
         letterSpacing: values.cfLetterSpacing,
         color: values.cfTextColor,
-        textAlign: values.cfTextAlign,
+        textAlign: values.cfTextAlign === 'left' ? 'start' : values.cfTextAlign === 'right' ? 'end' : values.cfTextAlign,
         textTransform: values.cfTextTransform,
         objectFit: values.cfImageOptions?.objectFit,
         objectPosition: values.cfImageOptions?.objectPosition,
